confluent_kafka/producer.py

```diff
@@ -0,0 +1,130 @@
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+
+        super(Producer, self).__init__(conf, error_cb=error_cb, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def produce(self, topic, value=None, key=None, partition=PARTITION_UA,
+                on_delivery=None, callback=None, timestamp=0, headers=None,
+                key_serializer=None, value_serializer=None):
+        """
+            Produces message to Kafka
+
+            :param str topic: Topic to produce message to.
+            :param str|bytes|obj value: Message payload; value_serializer required for non-character values.
+            :param str|bytes|obj key: Message key payload; key_serializer required for non-character keys.
+            :param int partition: Partition to produce to, else uses the configured built-in partitioner.
+                Default value is PARTITION_UA(round-robin over all partitions).
+            :param func on_delivery(confluent_kafka.KafkaError, confluent_kafka.Message):
+                Delivery report callback to call on successful or failed delivery.
+            :param func callback(confluent_kafka.KafkaError, confluent_kafka.Message):
+                See on_delivery.
+            :param int timestamp: Message timestamp (CreateTime) in milliseconds since epoch UTC
+                (requires librdkafka >= v0.9.4, api.version.request=true, and broker >= 0.10.0.0).
+                Default value is current time.
+            :param dict|list headers: Message headers to set on the message. The header key must be a string while
+                the value must be binary, unicode or None. Accepts a list of (key,value) or a dict.
+                (Requires librdkafka >= v0.11.4 and broker version >= 0.11.0.0)
+            :param func key_serializer(topic, key): Producer key_serializer override;
+                Converts message key to bytes. **note** serializers are responsible for handling NULL keys
+            :param func value_serializer(topic, value): Producer value_serializer override;
+                Converts message value to bytes. **note** serializers are responsible for handling NULL values
+
+        """
+
+        # on_delivery is an alias for callback and take precedence
+        if callback and not on_delivery:
+            on_delivery = callback
+
+        # parameter overrides take precedence over instance functions
+        if not key_serializer:
+            key_serializer = self._key_serializer
+
+        if key_serializer:
+            key = key_serializer(topic, key)
+
+        if not value_serializer:
+            value_serializer = self._value_serializer
+
+        if value_serializer:
+            value = value_serializer(topic, value)
+
+        super(Producer, self).produce(topic, value, key, partition, on_delivery=on_delivery,
```

Comment:
there is value in having the un-serialized values in the delivery reports, so i think this should be a feature of the library (though it has a cost so it should be togglable). however you're throwing this information away here.

Reply:
We don't have access to the message state from python which is one of the shortcomings of pulling this into python. What we can do however is provide several examples of delivery report callbacks which preserve the message contents pre-serialization which was the plan. Alternatively we can push this down into the extension, set the message contents with the msg key/value an avoid the extra work. Either one is doable although I think starting to pull things int python is nice. We could also wrap user provided callbacks in a lambda and change the docstring.

Reply:
this seems like a required feature to me. i didn't understand your plan with the extension - would you get a pointer to the original python value, pass that via the opaque parameter to the produce method, and get the value again using this pointer in the delivery report?

Reply:
We can call the serializer and hand off the message data in two places. 

1. In the extension, not the approach we took 
2. In python as we have done here. This means we do not have access to the message opaque pointer which is passed to produce however. 

If we went with step 1 we could simply provide the message state object with the data before calling serializer. Alternatively we could make a copy of the data and pass that to the message state object(which would be our opaque pointer). Either way we have access to the existing message state and the opaque pointer in the extension but not from python. 

For option 2 we can use a closure to bring the original message contents withit as we do in the avro-cli example. 

https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/avro-cli.py#L103

Reply:
@mhowlett I've thought about this more and I'm not sure I agree that providing the unserialized payload is the correct move here. If we do this then the delivery report is reporting what we *wanted* to produce as opposed to what we actually did produce. That to me breaks the contract that the user has with the delivery report. 

Using a lambda to carry the payload with your callback seems like the right move here. This ensures that the delivery report upholds its intended purpose while giving users the flexibility to cater to more specific use cases.

Reply:
There's also the possibly undesired or unexpected reference to the original user object that would be kept alive and out of gc until the delivery report was triggered, which may have unwanted bad consequences (resource usage, etc)